### PR TITLE
hasona23 (hm) Habit Logger project

### PR DESCRIPTION
**Pull request by hasona23**
dear sir,
this is the habit logger project

### CODE REVIEW

File: HabitTracker/HabitTracker/Program.cs [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+        Console.Clear();
+        Console.WriteLine("Enter habit's name or press enter to skip");
+        var name = "";
+        var date = "";
+        do
+        {
+            name = Console.ReadLine();
+        } while (!HabitDatabase.HabitTypes.ContainsKey(name) && !string.IsNullOrEmpty(name));
+
+
```
Comment (chrisjamiecarter) [inline HabitTracker/HabitTracker/Program.cs:54]: Code Conventions: Spacing. Redundant blank line (no double blank lines).

File: HabitTracker/HabitTracker/Program.cs [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+            var readLine = Console.ReadLine();
+
+            if (DateTime.TryParse(readLine, out var dateTime))
+            {
+                date = dateTime.ToString("yyyy-MM-dd");
+                break;
+            }
+        } while (true);
+
+        Console.WriteLine("Enter habit's amount ##-" + HabitDatabase.HabitTypes[name].Unit);
```
Comment (chrisjamiecarter) [inline HabitTracker/HabitTracker/Program.cs:107]: Use interpolated strings here:
❌ `Console.WriteLine("Enter habit's amount ##-" + HabitDatabase.HabitTypes[name].Unit);`
✔️ `Console.WriteLine($"Enter habit's amount ##-{HabitDatabase.HabitTypes[name].Unit}");`

File: HabitTracker/HabitTracker/Program.cs [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+            {
+                date = dateTime.ToString("yyyy-MM-dd");
+                break;
+            }
+        } while (true);
+
+        Console.WriteLine("Finally create the habit's measurement.");
+        Console.WriteLine("press enter to avoid or enter it in this format ##-unit");
+
+
```
Comment (chrisjamiecarter) [inline HabitTracker/HabitTracker/Program.cs:180]: Code Conventions: Spacing. Redundant blank line (no double blank lines).

File: HabitTracker/HabitTracker/Program.cs [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+        Console.WriteLine("press enter to avoid or enter it in this format ##-unit");
+
+        
+            var read = Console.ReadLine();
+            if (string.IsNullOrEmpty(read))
+                amount = null;
+            else
+                amount = read;
+        
+
```
Comment (chrisjamiecarter) [inline HabitTracker/HabitTracker/Program.cs:187]: Code Conventions: Spacing. Redundant blank line (no double blank lines).

File: HabitTracker/HabitTrackerLibrary/Habit.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+public class Habit
+{
+    private readonly DateTime _date;
+
+    public readonly string Name;
+
+    private readonly Amount? _amount;
+    //quantity is the optional amount to measure and displayed as "39-unit"
+
+
```
Comment (chrisjamiecarter) [inline HabitTracker/HabitTrackerLibrary/Habit.cs:12]: Code Conventions: Spacing. Redundant blank line (no double blank lines).

File: HabitTracker/HabitTrackerLibrary/Habit.cs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    {
+        return _date.ToString("yyyy-MM-dd");
+    }
+
+    public class Amount
+    {
+        public string Unit = "";
+        public int Value;
+
+        public Amount()
```
Comment (chrisjamiecarter) [inline HabitTracker/HabitTrackerLibrary/Habit.cs:42]: Is an empty constructor required here?

File: HabitTracker/HabitTrackerLibrary/HabitDatabase.cs [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+namespace HabitTrackerLibrary;
+
+public static class HabitDatabase
+{
+    private static readonly string _connectionString = "DataSource=HabitTracker.db";
+    public static Dictionary<string, Habit.Amount?> HabitTypes = new();
+
+    public static void CreateDatabase()
+    {
+        using (var connection = new SqliteConnection(_connectionString))
```
Comment (chrisjamiecarter) [inline HabitTracker/HabitTrackerLibrary/HabitDatabase.cs:12]: Remember that you can use the simpler "Using Declaration" syntax to save an indentation level against the "Using Statement" syntax:

❌ 
```
// Using Statement
using (var connection = new SqliteConnection(connectionString))
{
    // ...
}
```

✔️ 
```
// Using Declaration
using var connection = new SqliteConnection(connectionString);
// ...
```

File: README.md
Comment (chrisjamiecarter) [inline README.md:1]: This file needs to be a directory lower. Otherwise your project is not containerized and is mixed with other students.

**Review by chrisjamiecarter — CHANGES_REQUESTED** [review]
Hey @hasona23 👋,

Congratulations and great job on another project submission! 🎉 It’s great to see your progress.

I have performed a peer review and have some praise, as well as a few suggestions:
- Well written README.md file with great content! 🏅 
- Using File-Scoped Namespaces. 🎖️ 
- Implemented DRY principle. 🥇 
- Separation of concerns. ⭐ 
- Great code styling. 🌟 
- Perfect use of parameterised sql queries. 👏 

**Solution Structure**
Re-read step 3 of the [Code Reviews](https://thecsharpacademy.com/article/52/code-reviews) academy page.
When submitting your code review pull request, all your solution files should be contained inside a new folder as per below:

✔️ 
```
.
└── [ProjectName].[GitHubUserName]/
    ├── [SolutionDirectory1]/
    ├── [SolutionDirectory2]/
    └── [ProjectName].sln
```

❌ 
You are missing the `.[GitHubUserName]` suffix.
Your README.md file needs to be inside the above folder.
This will make it easier for code-reviewers to pull down and review your code in the future.

**Issues**
I have to double key some values. For example, on add new habit:

image

And on Delete:

image

When adding an existing habit there is not option to allow user to input Quantity:

image

The update function doesn't seem to work at all:

image

image

image

I suggest revisiting to try understand and fix any issues, as you are currently failing the below requirements:

- [ ] The users should be able to insert, delete, update and view their logged habit.
- [ ] You should handle all possible errors so that the application never crashes.

Feel free to reach out if you have any questions or if you'd like to collaborate further on any of these points.

You're doing a fantastic job, your code is good, there is just some logic issues that are easily fixed! Once resolved. Please commit and push your code. This will update me and I will finish my review!

Thanks,
@chrisjamiecarter

**Comment by hasona23** [comment]
Hello @chrisjamiecarter ,
thank your efforts reviewing the code i addressed all the issues except for the issues except for the update function the one where running doesnt have a unit to alter  and the double key 

Details:
double key issue:
the issue was caused due to a useless Console.ReadLine() in the DatabaseUpdateType function. 

Update Function malfunction:
the update function's aim is to just edit the habits quantity not the date and the error in it was that it resumed the execution when the function doesn't have any units to change which was due to a missed return; call to stop the execution of the functions which also resulted in the NULL REFRENCE error

program crashed :
It should be fixed now i stopped the execution of function and returned to the main menu when there is not quantity  so there should be no NULL REFRENCE error

**Review by chrisjamiecarter — APPROVED** [review]
Perfect @hasona23 🎉,

Well done on the updates.

I still think the user experience could be greatly improved but you have fulfilled all the requirements ⭐ 

I will go ahead and mark as approved, keep up the excellent work on the next projects! 😊

Best regards,
@chrisjamiecarter

**Pull request closed without merging**